### PR TITLE
[Improve](Nereids): refactor eliminate outer join

### PR DESCRIPTION
**Pull request by qzsee**
# Proposed changes

Refactor eliminate outer join #12985 

Evaluate the expression with `ConstantFoldRule`. If the evaluation result is `NULL` or `FALSE`, then the elimination condition is satisfied.

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221017114508_clickbench_pr_30005.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                    List<Expression> leftPredicates = Lists.newArrayList();
+                    List<Expression> rightPredicates = Lists.newArrayList();
+                    for (Expression predicate : ExpressionUtils.extractConjunction(filter.getPredicates())) {
+                        if (allCoveredBy(predicate, join.left().getOutputSet())) {
+                            leftPredicates.add(predicate);
+                        }
+                        if (allCoveredBy(predicate, join.right().getOutputSet())) {
+                            rightPredicates.add(predicate);
+                        }
+                    }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java:61]: nit: the current code abstraction level is not uniform, add a layer of abstraction for the detail of compute.

```java
List<Expression> leftPredicates = ExpressionUtils.extractCoveredConjunction(filter.getPredicates(), join.left().getOutputSet());
List<Expression> rightPredicates = ExpressionUtils.extractCoveredConjunction(filter.getPredicates(), join.right().getOutputSet());
```

keep in mind: rule should display the core logical firstly, and hide the compute detail in other methods or util class.

**Review by 924060929 — COMMENTED** [review]
good, looks more clearer

**Review by 924060929 — COMMENTED** [review]
good, looks more clearer

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalJoin())
+                .when(filter -> filter.child().getJoinType().isOuterJoin())
+                .then(filter -> {
+                    LogicalJoin<GroupPlan, GroupPlan> join = filter.child();
+                    List<Expression> leftPredicates = Lists.newArrayList();
+                    List<Expression> rightPredicates = Lists.newArrayList();
+                    for (Expression predicate : ExpressionUtils.extractConjunction(filter.getPredicates())) {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java:54]:
```suggestion
                    for (Expression predicate : filter.getConjuncts) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import com.google.common.collect.Maps;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Eliminate out join.
+ */
+public class EliminateOutJoin extends OneRewriteRuleFactory {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java:44]:
```suggestion
public class EliminateOuterJoin extends OneRewriteRuleFactory {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            Expression evalExpr = FoldConstantRule.INSTANCE.rewrite(ExpressionUtils.replace(predicate, replaceMap),
+                    new ExpressionRewriteContext(null));
+            if (nullLiteral.equals(evalExpr) || BooleanLiteral.FALSE.equals(evalExpr)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean allCoveredBy(Expression predicate, Set<Slot> inputSlotSet) {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java:106]:
```suggestion
    private boolean allCoveredBy(Expression predicate, Set<Slot> slots) {
```

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.27 seconds
 load time: 580 seconds
 storage size: 17154562461 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221019113657_clickbench_pr_30748.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuterJoin.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.Maps;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Eliminate out join.
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuterJoin.java:39]:
```suggestion
 * Eliminate outer join.
```

**Review by 924060929 — APPROVED** [review]

**Pull request merged**